Library/Source/NativeEngine.cpp

```diff
@@ -277,6 +320,155 @@ namespace babylon
+            float m_green{ 51.f / 255.f };
+            float m_blue{ 85.f / 255.f };
+            float m_alpha{ 1.f };
+            bool m_backBuffer{ true };
+            bool m_depth{ true };
+            bool m_stencil{ true };
+
+            uint32_t Color() const
+            {
+                uint32_t color = 0x0;
+                color += static_cast<uint8_t>(m_red * std::numeric_limits<uint8_t>::max());
+                color = color << 8;
+                color += static_cast<uint8_t>(m_green * std::numeric_limits<uint8_t>::max());
+                color = color << 8;
+                color += static_cast<uint8_t>(m_blue * std::numeric_limits<uint8_t>::max());
+                color = color << 8;
+                color += static_cast<uint8_t>(m_alpha * std::numeric_limits<uint8_t>::max());
+                return color;
+            }
+        };
+
+        struct FrameBufferData final
+        {
+            FrameBufferData(bgfx::FrameBufferHandle frameBuffer, RecycleSet<bgfx::ViewId>& viewIdSet, uint16_t width, uint16_t height)
+                : FrameBuffer{ frameBuffer }
+                , ViewId{ viewIdSet.Get() }
+                , ViewClearState{ ViewId }
+                , Width{ width }
+                , Height{ height }
+                , m_idSet{ viewIdSet }
+            {
+                assert(ViewId < bgfx::getCaps()->limits.maxViews);
+            }
+
+            FrameBufferData(FrameBufferData&) = delete;
+
+            ~FrameBufferData()
+            {
+                bgfx::destroy(FrameBuffer);
+                m_idSet.Recycle(ViewId);
+            }
+
+            void SetUpView()
+            {
+                bgfx::setViewFrameBuffer(ViewId, FrameBuffer);
+                ViewClearState.Update();
+                bgfx::setViewRect(ViewId, 0, 0, Width, Height);
+            }
+
+            bgfx::FrameBufferHandle FrameBuffer{ bgfx::kInvalidHandle };
+            bgfx::ViewId ViewId{};
+            ViewClearState ViewClearState;
+            uint16_t Width{};
+            uint16_t Height{};
+
+        private:
+            RecycleSet<bgfx::ViewId>& m_idSet;
+        };
+
+        struct FrameBufferManager final
+        {
+            FrameBufferData* CreateNew(bgfx::FrameBufferHandle frameBufferHandle, uint16_t width, uint16_t height)
+            {
+                return new FrameBufferData(frameBufferHandle, m_idSet, width, height);
+            }
+
+            void Bind(FrameBufferData* data)
+            {
+                assert(m_boundFrameBuffer == nullptr);
+                m_boundFrameBuffer = data;
+
+                // TODO: Consider doing this only on bgfx::reset(); the effects of this call don't survive reset, but as
+                // long as there's no reset this doesn't technically need to be called every time the frame buffer is bound.
+                m_boundFrameBuffer->SetUpView();
+
+                // bgfx::setTexture()? Why?
+                // TODO: View order?
```

Comment:
drawcalls can be reordered based on states, textures,... This is not a problem for a gbuffer with a depth buffer for solid geometry. But if you render things like GUI elements in a buffer with no depth buffer, this can be an issue.

Reply:
Spoke offline, added https://github.com/BabylonJS/BabylonNative/issues/42 to make sure ordering work doesn't get lost.

Reply:
Closed that issue, forgot I'd already added notes about these in #6 previously (https://github.com/BabylonJS/BabylonNative/pull/41#discussion_r319688614).